File: backend/app/schemas/listing.py
from typing import Annotated
from uuid import UUID, uuid4

from pydantic import BeforeValidator, Field, HttpUrl

from app.schemas.application import Application
from app.schemas.dates import ISODate
from app.schemas.status_event import StatusEnum
from app.schemas.types import CamelModel, parse_json_list_as


class ListingBase(CamelModel):
  title: str
  company: str
  domain: str
  location: str | None = None
  description: str
  posted_date: ISODate | None = None


class Listing(ListingBase):
  id: UUID = Field(default_factory=uuid4)
  url: HttpUrl

  skills: Annotated[
    list[str],
    BeforeValidator(parse_json_list_as(str)),
    Field(default_factory=list),
  ]

  requirements: Annotated[
    list[str],
    BeforeValidator(parse_json_list_as(str)),
    Field(default_factory=list),
  ]

  applications: list[Application] = Field(default_factory=list)


# TODO: DRY
class ListingSummary(CamelModel):
  id: UUID
  url: HttpUrl
  title: str
  company: str
  domain: str
  location: str | None = None
  posted_date: ISODate | None = None
  current_status: StatusEnum | None = None
  last_updated: str | None = None

File: frontend/src/types/listing.ts
import { ISODate } from '@/utils/date';

import type { Application, StatusEnum } from './application';

export type Listing = {
  id: string;
  url: string;
  title: string;
  company: string;
  domain: string;
  location: string | null;
  description: string;
  postedDate: ISODate | null;
  skills: string[];
  requirements: string[];
  applications: Application[];
};

export type ListingSummary = {
  id: string;
  url: string;
  title: string;
  company: string;
  domain: string;
  location: string | null;
  postedDate: ISODate | null;
  currentStatus: StatusEnum | null;
  lastUpdated: string | null;
};

export type GroundedItem<T = string> = {
  value: T;
  quote: string | null;
};

export type ListingExtraction = {
  title: string;
  company: string;
  domain: string;
  location: string | null;
  description: string;
  postedDate: ISODate | null;
  skills: GroundedItem<string>[];
  requirements: GroundedItem<string>[];
};

type BaseListingDraft = {
  id: string;
  url: string;
};

export type ListingDraftUnique = BaseListingDraft & {
  status: 'unique';
  listing: ListingExtraction;
  html: string | null;
};

export type ListingDraftDuplicateUrl = BaseListingDraft & {
  status: 'duplicate_url';
  duplicateOf: Listing;
  duplicateOfApplicationId: string;
};

export type ListingDraftDuplicateContent = BaseListingDraft & {
  status: 'duplicate_content';
  listing: ListingExtraction;
  duplicateOf: Listing;
  duplicateOfApplicationId: string;
  html: string | null;
};

export type ListingDraftError = BaseListingDraft & {
  status: 'error';
  error: string;
  html: string | null;
};

export type ListingDraftPending = BaseListingDraft & {
  status: 'pending';
};

export type ListingDraft =
  | ListingDraftPending
  | ListingDraftUnique
  | ListingDraftDuplicateUrl
  | ListingDraftDuplicateContent
  | ListingDraftError;
